File: CLAUDE/backend/src/services/zplPrinter.js
/**
 * Envio de ZPL para impressoras térmicas (Zebra/Elgin).
 * Porte de apps/api/src/infrastructure/zpl_printer.py.
 *
 * Modos:
 *   dry_run — grava o ZPL em arquivo local (teste sem hardware)
 *   raw     — TCP Raw Print (porta 9100 típica)
 *   cups    — `lp -d <fila> -o raw` (Linux/macOS com CUPS)
 */

const fs = require("fs");
const net = require("net");
const path = require("path");
const { execFile } = require("child_process");

const DRY_RUN_DIR =
  process.env.ZPL_DRY_RUN_DIR ||
  path.join(__dirname, "..", "..", "..", "var", "zpl_dry_run");

function resolvePrintMode(override) {
  const raw = String(override || process.env.ZPL_PRINT_MODE || "dry_run")
    .trim()
    .toLowerCase();
  if (["dry_run", "dry-run", "dryrun", "mock"].includes(raw)) return "dry_run";
  if (["raw", "socket", "9100"].includes(raw)) return "raw";
  if (["cups", "lp"].includes(raw)) return "cups";
  return "dry_run";
}

function printerStatus() {
  const mode = resolvePrintMode();
  const host = process.env.ZPL_PRINTER_HOST || "127.0.0.1";
  const port = Number(process.env.ZPL_PRINTER_PORT || 9100);
  const cupsQueue = process.env.ZPL_CUPS_QUEUE || "zebra";
  return {
    mode,
    host,
    port,
    cups_queue: cupsQueue,
    timeout_sec: Number(process.env.ZPL_PRINT_TIMEOUT_SEC || 8),
    dry_run_dir: DRY_RUN_DIR,
    ready_hint:
      mode === "dry_run"
        ? "Modo dry-run: bipagem grava ZPL em disco (sem impressora)."
        : mode === "raw"
        ? `Raw Print → ${host}:${port}`
        : `CUPS fila '${cupsQueue}' via lp -o raw`,
  };
}

function slugId(orderId) {
  return String(orderId || "sem-id")
    .replace(/[^A-Za-z0-9\-_]/g, "_")
    .slice(0, 80);
}

function printDryRun(zpl, orderId) {
  fs.mkdirSync(DRY_RUN_DIR, { recursive: true });
  const stamp = new Date().toISOString().replace(/[:.]/g, "").slice(0, 15);
  const file = path.join(DRY_RUN_DIR, `${stamp}_${slugId(orderId)}.zpl`);
  fs.writeFileSync(file, zpl, "utf8");
  fs.writeFileSync(path.join(DRY_RUN_DIR, "last_print.zpl"), zpl, "utf8");
  return {
    ok: true,
    mode: "dry_run",
    message: `Dry-run OK — ZPL salvo em ${path.basename(file)} (sem hardware).`,
    dry_run_path: file,
  };
}

function printRaw(zpl) {
  const host = process.env.ZPL_PRINTER_HOST || "127.0.0.1";
  const port = Number(process.env.ZPL_PRINTER_PORT || 9100);
  const timeout = Number(process.env.ZPL_PRINT_TIMEOUT_SEC || 8) * 1000;

  return new Promise((resolve) => {
    const socket = new net.Socket();
    let resolvido = false;
    const encerrar = (resultado) => {
      if (resolvido) return;
      resolvido = true;
      socket.destroy();
      resolve(resultado);
    };

    socket.setTimeout(timeout);
    socket.on("timeout", () =>
      encerrar({
        ok: false,
        mode: "raw",
        message: `Impressora não respondeu em ${timeout / 1000}s (${host}:${port}).`,
        detail: "timeout",
      })
    );
    socket.on("error", (err) =>
      encerrar({
        ok: false,
        mode: "raw",
        message: `Falha ao falar com a impressora ${host}:${port} — ${err.message}`,
        detail: err.code || "socket_error",
      })
    );
    socket.connect(port, host, () => {
      socket.write(zpl, "utf8", () =>
        encerrar({
          ok: true,
          mode: "raw",
          message: `Etiqueta enviada para ${host}:${port}.`,
        })
      );
    });
  });
}

function printCups(zpl, orderId) {
  const fila = process.env.ZPL_CUPS_QUEUE || "zebra";
  const timeout = Number(process.env.ZPL_PRINT_TIMEOUT_SEC || 8) * 1000;
  const tmp = path.join(
    fs.mkdtempSync(path.join(require("os").tmpdir(), "zpl-")),
    `${slugId(orderId)}.zpl`
  );
  fs.writeFileSync(tmp, zpl, "utf8");

  return new Promise((resolve) => {
    execFile("lp", ["-d", fila, "-o", "raw", tmp], { timeout }, (err, stdout, stderr) => {
      try {
        fs.unlinkSync(tmp);
      } catch (e) {
        /* arquivo temporário — ignorar */
      }
      if (err) {
        return resolve({
          ok: false,
          mode: "cups",
          message: `CUPS recusou o job na fila '${fila}': ${stderr || err.message}`,
          detail: String(err.code || ""),
        });
      }
      resolve({
        ok: true,
        mode: "cups",
        message: `Etiqueta enviada para a fila CUPS '${fila}'. ${String(stdout || "").trim()}`,
      });
    });
  });
}

/** Despacha o ZPL conforme o modo configurado. */
async function sendZpl(zpl, orderId, modeOverride) {
  const conteudo = String(zpl || "").trim();
  if (!conteudo) {
    return { ok: false, mode: resolvePrintMode(modeOverride), message: "ZPL vazio." };
  }
  const mode = resolvePrintMode(modeOverride);
  if (mode === "raw") return printRaw(conteudo);
  if (mode === "cups") return printCups(conteudo, orderId);
  return printDryRun(conteudo, orderId);
}

module.exports = { sendZpl, printerStatus, resolvePrintMode };

File: apps/api/src/infrastructure/zpl_printer.py
"""Envio de ZPL para impressoras térmicas (Zebra/Elgin).

Modos:
  - dry_run: grava o ZPL em arquivo local (teste sem hardware)
  - raw: TCP Raw Print (porta 9100 típica)
  - cups: `lp -d <fila> -o raw` (Linux/macOS com CUPS)
"""
from __future__ import annotations

import logging
import socket
import subprocess
import tempfile
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Literal, Optional

from src.config import settings

logger = logging.getLogger(__name__)

PrintMode = Literal["dry_run", "raw", "cups"]

DRY_RUN_DIR = Path(__file__).resolve().parent.parent.parent / "var" / "zpl_dry_run"


@dataclass
class PrintResult:
    ok: bool
    mode: str
    message: str
    detail: str = ""
    dry_run_path: Optional[str] = None

    def as_dict(self) -> dict:
        return {
            "ok": self.ok,
            "mode": self.mode,
            "message": self.message,
            "detail": self.detail,
            "dry_run_path": self.dry_run_path,
        }


def resolve_print_mode(override: Optional[str] = None) -> PrintMode:
    raw = (override or settings.ZPL_PRINT_MODE or "dry_run").strip().lower()
    if raw in ("dry_run", "dry-run", "dryrun", "mock"):
        return "dry_run"
    if raw in ("raw", "socket", "9100"):
        return "raw"
    if raw in ("cups", "lp"):
        return "cups"
    return "dry_run"


def printer_status() -> dict:
    mode = resolve_print_mode()
    return {
        "mode": mode,
        "host": settings.ZPL_PRINTER_HOST,
        "port": int(settings.ZPL_PRINTER_PORT),
        "cups_queue": settings.ZPL_CUPS_QUEUE,
        "timeout_sec": float(settings.ZPL_PRINT_TIMEOUT_SEC),
        "dry_run_dir": str(DRY_RUN_DIR),
        "ready_hint": (
            "Modo dry-run: bipagem grava ZPL em disco (sem impressora)."
            if mode == "dry_run"
            else (
                f"Raw Print → {settings.ZPL_PRINTER_HOST}:{settings.ZPL_PRINTER_PORT}"
                if mode == "raw"
                else f"CUPS fila '{settings.ZPL_CUPS_QUEUE}' via lp -o raw"
            )
        ),
    }


def _print_dry_run(zpl: str, order_id: str) -> PrintResult:
    DRY_RUN_DIR.mkdir(parents=True, exist_ok=True)
    stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    safe_id = "".join(c if c.isalnum() or c in "-_" else "_" for c in str(order_id))[:80]
    path = DRY_RUN_DIR / f"{stamp}_{safe_id}.zpl"
    path.write_text(zpl, encoding="utf-8")
    latest = DRY_RUN_DIR / "last_print.zpl"
    latest.write_text(zpl, encoding="utf-8")
    msg = f"Dry-run OK — ZPL salvo em {path.name} (sem hardware)."
    logger.info(msg)
    return PrintResult(ok=True, mode="dry_run", message=msg, dry_run_path=str(path))


def _print_raw(zpl: str) -> PrintResult:
    host = (settings.ZPL_PRINTER_HOST or "").strip()
    port = int(settings.ZPL_PRINTER_PORT or 9100)
    timeout = float(settings.ZPL_PRINT_TIMEOUT_SEC or 8.0)
    if not host:
        return PrintResult(
            ok=False,
            mode="raw",
            message="Impressora Raw não configurada: defina ZPL_PRINTER_HOST no .env.",
            detail="missing_host",
        )
    try:
        payload = zpl.encode("utf-8")
        with socket.create_connection((host, port), timeout=timeout) as sock:
            sock.settimeout(timeout)
            sock.sendall(payload)
        return PrintResult(
            ok=True,
            mode="raw",
            message=f"ZPL enviado via Raw Print para {host}:{port}.",
        )
    except OSError as exc:
        logger.warning("Raw print failed: %s", exc)
        return PrintResult(
            ok=False,
            mode="raw",
            message=(
                f"Falha ao falar com a impressora em {host}:{port}. "
                "Confira cabo/rede, se a Zebra/Elgin está ligada e se a porta 9100 está aberta."
            ),
            detail=str(exc),
        )


def _print_cups(zpl: str) -> PrintResult:
    queue = (settings.ZPL_CUPS_QUEUE or "").strip()
    if not queue:
        return PrintResult(
            ok=False,
            mode="cups",
            message="Fila CUPS não configurada: defina ZPL_CUPS_QUEUE no .env.",
            detail="missing_queue",
        )
    timeout = float(settings.ZPL_PRINT_TIMEOUT_SEC or 8.0)
    tmp_path: Optional[Path] = None
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", suffix=".zpl", delete=False, encoding="utf-8"
        ) as tmp:
            tmp.write(zpl)
            tmp_path = Path(tmp.name)
        proc = subprocess.run(
            ["lp", "-d", queue, "-o", "raw", str(tmp_path)],
            capture_output=True,
            text=True,
            timeout=timeout,
            check=False,
        )
        if proc.returncode != 0:
            err = (proc.stderr or proc.stdout or "lp falhou").strip()
            return PrintResult(
                ok=False,
                mode="cups",
                message=(
                    f"CUPS recusou a impressão na fila '{queue}'. "
                    "Verifique se o CUPS está ativo e se a Zebra/Elgin está cadastrada."
                ),
                detail=err,
            )
        return PrintResult(
            ok=True,
            mode="cups",
            message=f"ZPL enviado via CUPS para a fila '{queue}'.",
            detail=(proc.stdout or "").strip(),
        )
    except FileNotFoundError:
        return PrintResult(
            ok=False,
            mode="cups",
            message=(
                "Comando `lp` não encontrado neste sistema. "
                "Use ZPL_PRINT_MODE=raw (Windows/bancada) ou dry_run para testes."
            ),
            detail="lp_not_found",
        )
    except subprocess.TimeoutExpired:
        return PrintResult(
            ok=False,
            mode="cups",
            message=f"Timeout ao enviar para CUPS (fila '{queue}').",
            detail="timeout",
        )
    finally:
        if tmp_path and tmp_path.exists():
            try:
                tmp_path.unlink()
            except OSError:
                pass


def send_zpl(zpl: str, order_id: str = "", mode_override: Optional[str] = None) -> PrintResult:
    content = (zpl or "").strip()
    if not content:
        return PrintResult(
            ok=False,
            mode=resolve_print_mode(mode_override),
            message="ZPL vazio — nada a imprimir.",
            detail="empty_zpl",
        )
    mode = resolve_print_mode(mode_override)
    if mode == "dry_run":
        return _print_dry_run(content, order_id or "unknown")
    if mode == "raw":
        return _print_raw(content)
    return _print_cups(content)


SAMPLE_ZPL = """^XA
^CF0,40
^FO50,50^FDBASE ANTIGRAVITY - DRY RUN^FS
^FO50,110^FDPedido: {order_id}^FS
^FO50,170^FDEtiqueta engatilhada (Etapa 4)^FS
^FO50,230^BY2^BCN,80,Y,N,N^FD{barcode}^FS
^XZ
"""


def build_sample_zpl(order_id: str, barcode: str = "") -> str:
    code = barcode or order_id
    return SAMPLE_ZPL.format(order_id=order_id, barcode=code)
